src/lib/y2partitioner/dialogs/bcache.rb

```diff
@@ -0,0 +1,182 @@
+      def initialize(device_graph, device = nil)
+        # TODO: handle if device is passed
+        textdomain "storage"
+        @caching = CachingDevice.new(device_graph, device)
+        @backing = BackingDevice.new(device_graph, device, @caching)
+        @device_graph = device_graph
+      end
+
+      # @macro seeDialog
+      def title
+        _("Bcache Device")
+      end
+
+      # @macro seeDialog
+      def contents
+        VBox(
+          HBox(
+            @backing,
+            HSpacing(1),
+            @caching
+          ),
+          VSpacing(1)
+        )
+      end
+
+      def caching_device
+        @caching.result
+      end
+
+      def backing_device
+        @backing.result
+      end
+
+      # Widget to select the backing device
+      class BackingDevice < CWM::ComboBox
+        def initialize(device_graph, device, caching)
+          textdomain "storage"
+          @device = device
+          @device_graph = device_graph
+          @caching = caching
+        end
+
+        def label
+          _("Backing Device")
+        end
+
+        def items
+          available_devices.map do |dev|
+            [dev.name, dev.basename]
+          end
+        end
+
+        def help
+          # TODO: write it
+          ""
+        end
+
+        def init
+          return unless @device
+
+          self.value = @device.blk_device.name
+        end
+
+        def store
+          val = value
+          @result = @device_graph.blk_devices.find { |d| d.name == val }
+        end
+
+        def validate
+          if value.nil? || value.empty?
+            Yast2::Popup.show(
+              _("Empty backing device is not yet supported"),
+              headline: _("Cannot Create Bcache")
+            )
+            return false
+          elsif value == @caching.value
+            Yast2::Popup.show(
+              _("Backing and Caching device cannot be identical."),
+              headline: _("Cannot Create Bcache")
+            )
+            return false
+          else
+            true
+          end
+        end
+
+        attr_reader :result
+
+      private
+
+        def available_devices
+          @available ||= Y2Storage::Bcache.usable_blk_devices(@device_graph)
+        end
+      end
+
+      # Widget to select the caching device
+      class CachingDevice < CWM::ComboBox
+        def initialize(device_graph, device)
+          textdomain "storage"
+          @device_graph = device_graph
+          @device = device
+        end
+
+        def label
+          _("Caching Device")
+        end
+
+        def items
+          available_devices.map do |dev|
+            case dev
+            when Y2Storage::BcacheCset
+              # if set have more then 1 device,
+              # we still can identify it via this blk device as it cannot be shared
+              [dev.blk_devices.first.name, dev.display_name]
+            else
+              [dev.name, dev.basename]
```

Comment:
All devices in the devicegraph have an internal numeric identifier called `sid`. We never display it to the user, but you could use it as the item id, instead of using names.

Reply:
ok, so lets change it.